rfcs/records.md

```diff
@@ -0,0 +1,201 @@
+2. Type system compatibility. While Luau type checker can type tables, including ones with rigid structure, it falls short in OOP scenarios because it's very difficult
+to associate methods with table structure in idiomatic OOP in Luau with tables. Today the situation is especially dire because each method gets its own inferred self type
+(something that is likely to change), and it's impossible to specify a table-with-metatable type via obvious type syntax.
+3. Runtime overhead. While we heavily optimize table reads and writes in object-like scenario, there's still a non-zero cost that is paid for establishing the mapping
+from the index literal to the hash field. Method calls are similarly optimized but the optimization is restricted by the generality of table structure. Finally, large
+object size results in higher cache pressure which affects performance.
+4. Strictness of access. Today at runtime, reading an unknown key from the table returns `nil` and writing a new key just works. This creates significant amount of
+complexity in the type checker, as it has to differentiate between tables that are open for extension and closed via a set of heuristics, and results in easy to make mistakes in untyped code.
+
+Today, writing idiomatic objects in Luau is relatively straightforward but making idiomatic OOP type safe or maximally efficient is next to impossible.
+
+## Design
+
+This proposal suggests solving these problems with a new data type, called record. Record is a type that is semantically a dictionary with a fixed set of keys; much like
+a table, values associated with keys can be read or written to. Much like a table, it has a metatable that can be used to customize behavior of the object by providing
+extra operators (for arithmetics, stringification, etc.), as well as specifying methods.
+
+The difference between records and tables is that when creating a record, the VM allocates space for all values of all keys without having to duplicate the key data;
+since the structure of a record is immutable even if the contents isn't, this allows a much more efficient representation. A 6-field record will take estimated `8+32+6*16=136`
+bytes without implementation heroics, which is more than twice as efficient as table storage. For applications that use many objects this has potential to halve
+the memory footprint.
+
+Record fields use flexible types at runtime (we always allocate space for TValue and don't restrict writes into the record to a given type). In the future, we may
+introduce support for packed records where the table definition must use types and writes that don't abide by these types will trigger an error.
+
+> TODO: It's going to be difficult to migrate to packed records. Is there anything we can do right now to keep this possibility open without enforcing types at runtime?
```

Comment:
Is there much advantage in terms of efficiency for packed records? Like without `tt`, we're looking at 12 bytes for `Value` + `extra`. Would these pack tightly or would there be padding inserted?

Considering the generics section below, it seems preferable to me to keep types out of the question here. Since that avoids the need for any kind of monomorphization, the field types can be erased at runtime.

Reply:
The thought is that we'd omit extra as well when we know that the type isn't a vector. But yeah I forgot about generics and only remembered them later.

Reply:
Yeah, that's the same restriction as we have today for all other types of data storage (stacks, array entries, table entries)